nova_vm/src/ecmascript/abstract_operations/operations_on_iterator_objects.rs

```diff
@@ -0,0 +1,327 @@
+                agent,
+                obj,
+                PropertyKey::Symbol(WellKnownSymbolIndexes::Iterator.into()),
+            )?
+            else {
+                // ii. If syncMethod is undefined, throw a TypeError exception.
+                return Err(
+                    agent.throw_exception(ExceptionType::TypeError, "No iterator on object")
+                );
+            };
+
+            // iii. Let syncIteratorRecord be ? GetIteratorFromMethod(obj, syncMethod).
+            let sync_iterator_record = get_iterator_from_method(agent, obj, sync_method)?;
+
+            // iv. Return CreateAsyncFromSyncIterator(syncIteratorRecord).
+            todo!("Implement create_async_from_sync_iterator(sync_iterator_record)")
+        } else {
+            method
+        }
+    } else {
+        // 2. Else,
+        // a. Let method be ? GetMethod(obj, @@iterator).
+        get_method(
+            agent,
+            obj,
+            PropertyKey::Symbol(WellKnownSymbolIndexes::Iterator.into()),
+        )?
+    };
+
+    // 3. If method is undefined, throw a TypeError exception.
+    let Some(method) = method else {
+        return Err(agent.throw_exception(
+            ExceptionType::TypeError,
+            "Iterator method cannot be undefined",
+        ));
+    };
+
+    // 4. Return ? GetIteratorFromMethod(obj, method).
+    get_iterator_from_method(agent, obj, method)
+}
+
+/// [7.4.4 IteratorNext ( iteratorRecord [ , value ] )](https://tc39.es/ecma262/#sec-iteratornext)
+///
+/// The abstract operation IteratorNext takes argument iteratorRecord (an
+/// Iterator Record) and optional argument value (an ECMAScript language value)
+/// and returns either a normal completion containing an Object or a throw
+/// completion.
+pub(crate) fn iterator_next(
+    agent: &mut Agent,
+    iterator_record: &IteratorRecord,
+    value: Option<Value>,
+) -> JsResult<Object> {
+    // 1. If value is not present, then
+    // a. Let result be ? Call(iteratorRecord.[[NextMethod]], iteratorRecord.[[Iterator]]).
+    // 2. Else,
+    // a. Let result be ? Call(iteratorRecord.[[NextMethod]], iteratorRecord.[[Iterator]], « value »).
+    let result = call(
+        agent,
+        iterator_record.next_method,
+        iterator_record.iterator.into(),
+        value.as_ref().map(|v| std::slice::from_ref(v)),
+    )?;
+
+    // 3. If result is not an Object, throw a TypeError exception.
+    // 4. Return result.
+    result.try_into().or(Err(agent.throw_exception(
+        ExceptionType::TypeError,
+        "The iterator result was not an object",
+    )))
+}
+
+/// [7.4.5 IteratorComplete ( iterResult )](https://tc39.es/ecma262/#sec-iteratorcomplete)
+///
+/// The abstract operation IteratorComplete takes argument iterResult (an
+/// Object) and returns either a normal completion containing a Boolean or a
+/// throw completion.
+pub(crate) fn iterator_complete(agent: &mut Agent, iter_result: Object) -> JsResult<Value> {
```

Comment:
**nitpick:** Not your fault obviously since the return value type comes from `to_boolean` but this should probably be `JsResult<bool>`.

Reply:
Yeah leaving this for the future then.